### ISSUE TITLE
[SIP-93] Proposal for non-blocking SQL Lab persistence

### ISSUE THREAD

**Issue by justinpark** [apache/superset#21385]
### Motivation

At Airbnb we have a performance (lag) issue using `SQLLAB_BACKEND_PERSISTENCE` feature.
Since `SQLLAB_BACKEND_PERSISTENCE` designed in a blocking way (as shown in the following), user can be stuck while `tab_state` table in busy state. (even unable to run a query while the `tabstateview` endpoint timed out)

https://github.com/apache/superset/blob/8c2719b11cebe451c5ba68193cf8dd51e4dce8e3/superset-frontend/src/SqlLab/actions/sqlLab.js#L614-L625

In order to improve this lag issue, we'd like to have a non-blocking mechanism for this db syncing workflow. i.e. like g-doc auto saving. moreover we'd like to design a hybrid solution (i.e. keep using redux-localstorage and save same snaphot in remote) to keep the advantage (offline saving) of using local storage.

### Proposed Change

The proposed solution is simple. We would like to keep using redux-localstorage as well as the current localStorage mechanism (as SQLLAB_BACKEND_PERSISTENCE turned OFF). Like QueryAutoSync does, we would like to build a component that observes the changes of queryEditor and then periodically sync the updated queryEditor snapshot from redux state using existing `/tabstateview/` api.
For the payload wise, we'd like to use `extra_json` in `tab_state` table to store the snapshot(by json format) of queryEditor as stored in localStorage (rather than mapping each value in a proper table column like sql, label, query_limit).
As existing persistence feature does, we will design similar hydration flow in frontend to initialize the sqlLab state from the backend. With this json blob it will simply convert to json and then merge with the existing local state.

### New or Changed Public Interfaces

We will replace prefetch(`/tabstateview/` call) block code in sqllab actions on SQLLAB_BACKEND_PERSISTENCE (~~or introduce new `SQLLAB_BACKEND_PERSISTENCE_ASYNC` FLAG if preferred~~) and need to pass the `active_tab` value including `extra_json` in this case. 

### New dependencies

I do not for-see any new dependencies.

### Migration Plan and Compatibility

I do not for-see any migration required.

### Rejected Alternatives

None.

to: @etr2460 @ktmud @john-bodley
cc: @betodealmeida

**Comment by ktmud** [comment]
I like this idea. Can you elaborate more on which states will be pulled to local storage and which states will be kept exclusively on remote? I'd imagine we probably shouldn't keep even a temporary local copy of big payloads such as query results.

<img width="485" alt="image" src="https://user-images.githubusercontent.com/335541/189205958-31f8737e-ad17-4331-af85-a8d24f4b7061.png">


For states that can be loaded fresh upon page load--such as `tables` and `databases`, they should also be skipped for localStorage in this async approach.

<img width="425" alt="image" src="https://user-images.githubusercontent.com/335541/189204207-ef8235a4-e979-4b17-b644-7fab0cadf901.png">

There are also some duplicates in saved states, too, that we may want to clean up:

<img width="474" alt="image" src="https://user-images.githubusercontent.com/335541/189204753-f8853ae5-5fdb-493d-9c32-00bc55b33e0b.png">

I also noticed we are saving previous queries of current active tab even though they may never be used again. Should be able to save a lot of space if we just clean them up.

---

To summarize, I think it'd be useful to start cleaning up current saved states when `SQLLAB_BACKEND_PERSISTENCE` is off, before migrating to this hybrid approach.

**Comment by justinpark** [comment]
> I like this idea. Can you elaborate more on which states will be pulled to local storage and which states will be kept exclusively on remote? I'd imagine we probably shouldn't keep even a temporary local copy of big payloads such as query results.
> 

Good point. There's a helper function for this and planning to reuse this function to trim the unnecessary items.

https://github.com/apache/superset/blob/5f9f657805a5559de6f06c647a33f278aadc32db/superset-frontend/src/SqlLab/utils/reduxStateToLocalStorageHelper.js#L56-L60

> To summarize, I think it'd be useful to start cleaning up current saved states 

100% agree. we should clean up the following keys and save only necessary

https://github.com/apache/superset/blob/5f9f657805a5559de6f06c647a33f278aadc32db/superset-frontend/src/SqlLab/utils/reduxStateToLocalStorageHelper.js#L21-L37

**Comment by geido** [comment]
Nice proposal @justinpark

**Comment by mistercrunch** [comment]
Took a look today, this is comprehensive, seems reasonable and I don't think there's anything controversial here.

+1

Though I'd love to get @betodealmeida 's blessing since I believe he implemented the original `SQLLAB_BACKEND_PERSISTENCE`

**Comment by justinpark** [comment]
> If it's the latter, we could reuse the same feature flag given that we're planning to clean up the current state, which will modify the existing feature anyway.

@michael-s-molina 
You're right but I suggested new feature flag to reduce the failure/loss of existing persist functionality due to the error by new integration.

I hope to make new async persistence option first and then replace/consolidate with the same persist feature flag name once no running bugs found from our cases.

**Comment by michael-s-molina** [comment]
Closing the SIP as it was approved.
